Automattic/pocket-casts-ios
# Up Next Queue Syncing: Attempt to merge the local and remote device changes to prevent items from the queue being removed or cleared

Part of #24

> **Note**: The majority of the line changes are actually me adding a ton of documentation and logging. 

## Description

This attempts to solve the instances where a users up next queue can be incorrectly cleared, or replaced with the contents of an outdated account. For more info on the scenarios that this could occur [see this comment](https://github.com/Automattic/pocket-casts-ios/issues/24#issuecomment-1397858597).

In order to prevent a users queue from being removed we now apply different logic based on the context ("reason") that a sync is occurring.

#### Account Creation
If we are syncing because due to a new account being created, we treat the local queue as the source of truth and it is then persisted to the server. 

#### Logging In
If we are syncing because the user is logging into an existing account we apply a non-destructive merge:
- Retrieve the latest up next queue from the server
- Add all the remote items to the users local queue
- Ensure any local episodes that were not on the server are not removed
- Persist the "merged" changes to the server

#### Everything else 
We apply the same logic as we did before. 

### Known issues with this approach

#### Removed episodes reappearing
Since during a login we only add items to the queue, and never remove them a user who has previously removed episodes from their queue may have them added back in if they login on a device that is not up to date. 

However, I think this is a worthy trade off to deleting or replacing users queues entirely. 

#### Sort order not persisting for logged out users

During a login the latest server changes are applied to the episodes in the queue which means any local episodes that were added may appear out of their original order. 

Again, I think this is an okay trade off. 

## Demo Videos

<details>
  <summary><h3>Installing the app from scratch and logging into an existing account that has a queue</h3>
</summary>

---
> ✅ **Expectation**: The existing users queued will be persisted on the new installation.
---

https://user-images.githubusercontent.com/793774/213831326-3cb5fa0a-98b2-4cf7-ab6a-13809f67a1a4.mov

</details>

<details>
<summary><h3>Logged out user with queued items, logging into existing account with a queue</h3></summary>

---
> ✅ **Expectation**: The logged out and existing queued items will be merged.
---

https://user-images.githubusercontent.com/793774/213831680-07b4e2b2-622c-4c71-b1ec-4d5894f4a679.mov

</details>

<details>
<summary><h3>Logged out user with queued items, logging into an account that has no queue</h3></summary>

---
> ✅ **Expectation**: The logged out queued items will appear on the account
---

https://user-images.githubusercontent.com/793774/213831935-83edc9af-9ce9-4e8d-9130-b758ba93023c.mov

</details>

<details>
<summary><h3>Logged out user with items that were removed on the other device</h3></summary>

---
> ✅ **Expectation**: The deleted items will be restored when the logged out device logs back in 
---

https://user-images.githubusercontent.com/793774/213832139-dbc25743-8ec7-495f-bbe8-7915d97927db.mov

</details>

## To test

### Account Creation

1. Reinstall the app
2. Dismiss the login prompt
3. Add some items to your up next queue
4. Create a new account
5. ✅ Verify your local queue remains the same as it was
6. Sign into the new account on another device
7. ✅ Verify the correct queue is sync'd

### Installing the app from scratch and logging into an existing account that has a queue

1. Sign into an account on device 1
2. Add some items to your up next queue and sync them
3. Reinstall the app on device 2
4. When prompted tap login
5. Sign into the device 1 account
6. ✅ Verify the correct queue is added to your device
7. Go to Device 1, and perform a sync
8. ✅ Verify the queue remains the same 

### Being a logged out user with items added to the up next, and logging into an account that has no items in the queue

1. Sign into an account on device 1
2. Clear your queue
3. Reinstall the app on device 2
4. Dismiss the login prompt
5. Add some items to the queue
7. Sign into the device 1 account
8. ✅ Verify your local queue remains the same
9. Go to Device 1, and perform a sync
10. ✅ Verify the device 2 queue is moved over

### Being a logged out user with items added to the up next, and logging into an account that has existing items in the queue

1. Sign into an account on device 1
2. Add some items to your up next queue and sync them
3. Reinstall the app on device 2
4. Dismiss the login prompt
5. Add some items to the queue
7. Sign into the device 1 account
8. ✅ Verify the queue consists of the items you added and the items from device 1
9. Go to Device 1, and perform a sync
10. ✅ Verify you see both device 1 and 2's queue

### Being a logged out user who removed items from their queue, and logging into an account where those changes weren't persisted

1. Sign into an account on device 1
2. Add some items to your up next queue and sync them
3. Reinstall the app on device 2
4. Sign into device 1's account
5. Then sign out
8. Remove some items from the queue
9. Sign back into device 1's account
10. ✅ ⚠️ The items that you removed will be added back to device 2 (See known issues above)

### Being a logged in user both devices and syncing changes
1. Sign into an account on device 1
2. Add some items to your up next queue and sync them
4. On Device 2, Sign into device 1's account
5. On Device 1, make some changes to the queue (add, remove, move, etc)
8. Sync the changes
9. On device 2, refresh
10. ✅ The queue is updated correctly
11. On device 2, make some changes to the queue. 
12. Sync them
13. On device 1, refresh
14. ✅ Verify the changes are sync'd back correctly.


## Checklist
